gst/nnstreamer/include/nnstreamer_plugin_api_filter.h

```diff
@@ -172,6 +172,8 @@ typedef struct _GstTensorFilterFrameworkInfo
   int verify_model_path; /**< TRUE(nonzero) when the NNS framework, not the sub-plugin, should verify the path of model files. */
   const accl_hw *hw_list; /**< List of supported hardwares by the framework.  Positive response of this check does not guarantee successful running of model with this accelerator. Subplugin is supposed to allocate/deallocate. */
   int num_hw; /**< number of hardware accelerators in the hw_list supported by the framework */
+  accl_hw accl_auto;  /**< accelerator to be used in auto mode (acceleration to be used but accelerator is not specified for the filter) - default -1 implies use first entry from hw_list */
+  accl_hw accl_default;   /**< accelerator to be used by default (valid user input is not provided) - default -1 implies use first entry from hw_list*/
```

Comment:
Question: can each subplugin alter auto/default values in run-time and affect the behaviors?
If then, LGTM.

Reply:
If the plugin is closed, then yes, its effect will be reflected in run-time (opening the plugin would call `getFrameworkInfo()` which will get updated values from the framework itself).
If the plugin is already opened, then only reopen will reflect the update in runtime.

But this is all assuming that the filter returns the updated value with `getFrameworkInfo()`.

Reply:
If the subplugin is already opened, then a reopen can update the filter.
If we want to trigger the change without a reopen, the selected ACCL (based on the user-input) can be made available with `GstTensorFilterProperty` to the plugin, and the plugin needs to do it by itself.

For example - if user input it `true:auto` or `true:auto,gpu,cpu` etc, the `GstTensorFilterProperty` will contain info that accl `auto` is selected (when opening the plugin, this information is not passed. Rather, just the specified accelerator designated is `auto` is passed). Then when the plugin updates its auto/default values, it can check that the `auto` value is being used, and the plugin should by itself update its behaviors.
This change is not being driven by `tensor_filter` as subplugin cannot notify tensor_filter that it has altered any of its properties/configuration (`tensor_filter` always starts the communication between itself and its subplugins).

However, note that this will not support updating the supported accelerator lists because that would require matching it with the user-provided input which is hidden from the subpluins.

Reply:
If a filter's option is auto and it may use NPU-1 and NPU-2, we may want to use different NPU in run-time (e.g., if NPU-1's utilization by another process is too high, use NPU-2)

We do not have such filters prepared, but in the future, I expect that users will want to use "auto" for such purposes. For now, we just need to "allow" a subplugin to do so. Note that this does not necessarily require yet another callback, we may assume that a subplugin may change it in a invoke callback while a subplugin who does not aware of "auto" or "default" will get one (other than auto/default) automatically). **Anyway.. this looks self-contradicting...**: we cannot achieve both simultaneously in the current structure of nnstreamer. :(

So.. for now... your approach looks fair enough.

LGTM.

Reply:
> If a filter's option is auto and it may use NPU-1 and NPU-2, we may want to use different NPU in run-time (e.g., if NPU-1's utilization by another process is too high, use NPU-2)

In this example, switching between NPU can be done by two entities - 
1. By subplugin itself - this can be done by using the hierarchy in the accl enum. NPU enum would include NPU-1 and NPU-2. NPU-1 enum only marks NPU-1 and same for NPU-2 (if NPU-1 and NPU-2 have the same common enum NPU, then also this applies). If filter auto sets NPU enum, then filter is opened with NPU enum, giving it anonymity to choose which NPU to use and update it in run-time based on utilization or other criteria.
2. By nnstreamer itself - other than scheduling required to be done in nnstreamer, passing just a device enum with every invoke can do the job (this will change the internal APIs again though).  
 
> We do not have such filters prepared, but in the future, I expect that users will want to use "auto" for such purposes. For now, we just need to "allow" a subplugin to do so. 

For this, I will add the property to `GstTensorFilterProperty` as mentioned in https://github.com/nnstreamer/nnstreamer/pull/2372#discussion_r424939871.